Index: smtp.module
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/smtp/smtp.module,v
retrieving revision 1.17.2.12
diff -u -r1.17.2.12 smtp.module
--- smtp.module	10 Nov 2009 17:03:10 -0000	1.17.2.12
+++ smtp.module	27 Feb 2010 17:33:44 -0000
@@ -397,8 +397,11 @@
             $content_type = 'text/html';
           break;
           case 'multipart/related':
+            // Get the boundary ID from the Content-Type header.
+            $boundary = _smtp_get_substring($value, 'boundary', '"', '"');
+            
             // The message includes an HTML part w/inline attachments.
-            $mail->ContentType = $content_type = 'multipart/related';
+            $mail->ContentType = $content_type = 'multipart/related; boundary="'. $boundary .'"';
           break;
           case 'multipart/alternative':
             // The message includes both a plain text and an HTML part.
